dm/pkg/conn/utils.go

```diff
@@ -144,23 +141,63 @@ func GetMasterStatus(ctx *tcontext.Context, db *BaseDB, flavor string) (
 		| 0-1-2                    |
 		+--------------------------+
 	*/
+
+	var rowsResult [][]string
 	if flavor == gmysql.MySQLFlavor {
-		err = rows.Scan(&binlogName, &pos, &binlogDoDB, &binlogIgnoreDB, &gtidStr)
+		rowsResult, err = export.GetSpecifiedColumnValuesAndClose(rows, "File", "Position", "Binlog_Do_DB", "Binlog_Ignore_DB", "Executed_Gtid_Set")
+		if err != nil {
+			err = terror.DBErrorAdapt(err, terror.ErrDBDriverError)
+			return binlogName, pos, binlogDoDB, binlogIgnoreDB, gtidStr, err
+		}
+
+		if len(rowsResult) == 0 {
+			err = terror.ErrNoMasterStatus.Generate()
+			return binlogName, pos, binlogDoDB, binlogIgnoreDB, gtidStr, err
+		} else if len(rowsResult[0]) != 0 {
+			binlogName = rowsResult[0][0]
+			var posInt int
+			posInt, err = strconv.Atoi(rowsResult[0][1])
+			if err != nil {
+				err = terror.DBErrorAdapt(err, terror.ErrDBDriverError)
+				return binlogName, pos, binlogDoDB, binlogIgnoreDB, gtidStr, err
+			}
+			pos = uint32(posInt)
```

Comment:
convert int to uint32 is a bit dangerous because int type can be 32 or 64 bits.
We can parse it to type of 64 bit first and then convert it to uint32

and suddenly I found this should be int64 😂  https://github.com/mysql/mysql-server/blob/a246bad76b9271cb4333634e954040a970222e0a/sql/rpl_source.cc#L1225-L1226 we can open an issue to record

Reply:
Modified